xrdcp fails when redirected for current versions

Hi all,
We currently have problems with xrdcp for our local resources.
We have a XRootD server and redictor on a ceph node to provide local datasets of our users to the (external) worker nodes of our T3 via XRootD.

**Setup**:
Ceph configs: [data server](https://etpwww.etp.kit.edu/~rhofsaess/xrootd-ceph-ds.cfg),  [redirector](https://etpwww.etp.kit.edu/~rhofsaess/xrootd-ceph-mgr.cfg)

On the worker node, a caching proxy and a redirector are running. Configs: [proxy](https://etpwww.etp.kit.edu/~rhofsaess/xrootd-cache-server.cfg), [redirector](https://etpwww.etp.kit.edu/~rhofsaess/xrootd-cache-manager.cfg)

The client is configured to use the redirector. Config: [client-plugin-proxy.conf](https://etpwww.etp.kit.edu/~rhofsaess/client-plugin-proxy.conf)

**What we observe**
xrdcp fails for current versions (tested with: 5.5.1)

What we see is that the final url is build up correctly (`final_url=root://f03-001-110-e.gridka.de:1094//root://ceph-node-j.etp.kit.edu:1094//mschnepf/data_mnist-tar.gz`), but the transfer somewhat fails with the error:

```
[0B/0B][100%][==================================================][0B/s]
Run: [ERROR] Operation is not implemented:  (source)
```
Full log: [transfer_5-5-1.log](https://etpwww.etp.kit.edu/~rhofsaess/transfer_5-5-1.log)

With older versions of xrootd, however, our setup works fine, see [transfer_5-3-1.log](https://etpwww.etp.kit.edu/~rhofsaess/transfer_5-3-1.log).

I tracked it down to the client(config) or the redirector.
If I do not use the redirector but ask the server directly, it works with a current xrootd version (5.5.4). [Log: transfer_5-5-4-without.log](https://etpwww.etp.kit.edu/~rhofsaess/transfer_5-5-4-without.log)

What I also observed was that changing the major version in "lib =libXrdClProxyPlugin-<V>.so" in the client plugin conf
does not have any influence. When changing it to V=1000, the redirect still works (and it still fails :D).
Just wanted to make sure that this behaviour is intended.

Maybe also our configs may aren't compatible with newer versions. If so, thanks for any help in advance to fix it! :)

Thanks and cheers

--- a/src/XrdApps/XrdClProxyPlugin/ProxyPrefixFile.hh
+++ b/src/XrdApps/XrdClProxyPlugin/ProxyPrefixFile.hh
@@ -26,6 +26,8 @@
 #include "XrdCl/XrdClDefaultEnv.hh"
 #include "XrdCl/XrdClPlugInInterface.hh"
 
+#include <utility>
+
 using namespace XrdCl;
 
 namespace xrdcl_proxy
@@ -45,7 +47,7 @@ public:
   //----------------------------------------------------------------------------
   //! Destructor
   //----------------------------------------------------------------------------
-  virtual ~ProxyPrefixFile();
+  virtual ~ProxyPrefixFile() override;
 
   //----------------------------------------------------------------------------
   //! Open
@@ -54,13 +56,13 @@ public:
                             OpenFlags::Flags flags,
                             Access::Mode mode,
                             ResponseHandler* handler,
-                            uint16_t timeout);
+                            uint16_t timeout) override;
 
   //----------------------------------------------------------------------------
   //! Close
   //----------------------------------------------------------------------------
   virtual XRootDStatus Close(ResponseHandler* handler,
-                             uint16_t         timeout)
+                             uint16_t         timeout) override
   {
     return pFile->Close(handler, timeout);
   }
@@ -70,7 +72,7 @@ public:
   //----------------------------------------------------------------------------
   virtual XRootDStatus Stat(bool             force,
                             ResponseHandler* handler,
-                            uint16_t         timeout)
+                            uint16_t         timeout) override
   {
     return pFile->Stat(force, handler, timeout);
   }
@@ -83,9 +85,21 @@ public:
                             uint32_t         size,
                             void*            buffer,
                             ResponseHandler* handler,
-                            uint16_t         timeout)
+                            uint16_t         timeout) override
   {
     return pFile->Read(offset, size, buffer, handler, timeout);
+  }
+
+  //------------------------------------------------------------------------
+  //! PgRead
+  //------------------------------------------------------------------------
+  virtual XRootDStatus PgRead( uint64_t         offset,
+                               uint32_t         size,
+                               void            *buffer,
+                               ResponseHandler *handler,
+                               uint16_t         timeout ) override
+  {
+    return pFile->PgRead(offset, size, buffer, handler, timeout);
   }
 
   //----------------------------------------------------------------------------
@@ -95,16 +109,53 @@ public:
                              uint32_t         size,
                              const void*      buffer,
                              ResponseHandler* handler,
-                             uint16_t         timeout)
+                             uint16_t         timeout) override
   {
     return pFile->Write(offset, size, buffer, handler, timeout);
+  }
+
+  //------------------------------------------------------------------------
+  //! Write
+  //------------------------------------------------------------------------
+  virtual XRootDStatus Write( uint64_t          offset,
+                              Buffer          &&buffer,
+                              ResponseHandler  *handler,
+                              uint16_t          timeout = 0 ) override
+  {
+    return pFile->Write(offset, std::move(buffer), handler, timeout);
+  }
+
+  //------------------------------------------------------------------------
+  //! Write
+  //------------------------------------------------------------------------
+  virtual XRootDStatus Write( uint64_t            offset,
+                              uint32_t            size,
+                              Optional<uint64_t>  fdoff,
+                              int                 fd,
+                              ResponseHandler    *handler,
+                              uint16_t            timeout = 0 ) override
+  {
+    return pFile->Write(offset, size, fdoff, fd, handler, timeout);
+  }
+
+  //------------------------------------------------------------------------
+  //! PgWrite
+  //------------------------------------------------------------------------
+  virtual XRootDStatus PgWrite( uint64_t               offset,
+                                uint32_t               nbpgs,
+                                const void            *buffer,
+                                std::vector<uint32_t> &cksums,
+                                ResponseHandler       *handler,
+                                uint16_t               timeout ) override
+  {
+    return pFile->PgWrite(offset, nbpgs, buffer, cksums, handler, timeout);
   }
 
   //----------------------------------------------------------------------------
   //! Sync
   //----------------------------------------------------------------------------
   virtual XRootDStatus Sync(ResponseHandler* handler,
-                            uint16_t         timeout)
+                            uint16_t         timeout) override
   {
     return pFile->Sync(handler, timeout);
   }
@@ -114,7 +165,7 @@ public:
   //----------------------------------------------------------------------------
   virtual XRootDStatus Truncate(uint64_t         size,
                                 ResponseHandler* handler,
-                                uint16_t         timeout)
+                                uint16_t         timeout) override
   {
     return pFile->Truncate(size, handler, timeout);
   }
@@ -125,9 +176,31 @@ public:
   virtual XRootDStatus VectorRead(const ChunkList& chunks,
                                   void*            buffer,
                                   ResponseHandler* handler,
-                                  uint16_t         timeout)
+                                  uint16_t         timeout) override
   {
     return pFile->VectorRead(chunks, buffer, handler, timeout);
+  }
+
+  //------------------------------------------------------------------------
+  //! VectorWrite
+  //------------------------------------------------------------------------
+  virtual XRootDStatus VectorWrite( const ChunkList &chunks,
+                                    ResponseHandler *handler,
+                                    uint16_t         timeout = 0 ) override
+  {
+    return pFile->VectorWrite(chunks, handler, timeout);
+  }
+
+  //------------------------------------------------------------------------
+  //! @see XrdCl::File::WriteV
+  //------------------------------------------------------------------------
+  virtual XRootDStatus WriteV( uint64_t            offset,
+                               const struct iovec *iov,
+                               int                 iovcnt,
+                               ResponseHandler    *handler,
+                               uint16_t            timeout = 0 ) override
+  {
+    return pFile->WriteV(offset, iov, iovcnt, handler, timeout);
   }
 
   //----------------------------------------------------------------------------
@@ -135,7 +208,7 @@ public:
   //----------------------------------------------------------------------------
   virtual XRootDStatus Fcntl(const Buffer&    arg,
                              ResponseHandler* handler,
-                             uint16_t         timeout)
+                             uint16_t         timeout) override
   {
     return pFile->Fcntl(arg, handler, timeout);
   }
@@ -144,7 +217,7 @@ public:
   //! Visa
   //----------------------------------------------------------------------------
   virtual XRootDStatus Visa(ResponseHandler* handler,
-                            uint16_t         timeout)
+                            uint16_t         timeout) override
   {
     return pFile->Visa(handler, timeout);
   }
@@ -152,7 +225,7 @@ public:
   //----------------------------------------------------------------------------
   //! IsOpen
   //----------------------------------------------------------------------------
-  virtual bool IsOpen() const
+  virtual bool IsOpen() const override
   {
     return pFile->IsOpen();
   }
@@ -161,7 +234,7 @@ public:
   //! SetProperty
   //----------------------------------------------------------------------------
   virtual bool SetProperty(const std::string& name,
-                           const std::string& value)
+                           const std::string& value) override
   {
     return pFile->SetProperty(name, value);
   }
@@ -170,7 +243,7 @@ public:
   //! GetProperty
   //----------------------------------------------------------------------------
   virtual bool GetProperty(const std::string& name,
-                           std::string& value) const
+                           std::string& value) const override
   {
     return pFile->GetProperty(name, value);
   }

